### PR TITLE
[FLINK-15482][hive] Failed to call Hive functions returning decimal type

### PR DESCRIPTION
**Pull request by lirui-apache**
## What is the purpose of the change

Fix the issue of calling Hive functions that return decimals

## Brief change log

  - Call `TypeInfoDataTypeConverter` to convert data type to type info because it's the recommended way.
  - Add test case

## Verifying this change

New test case

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? NA

### CODE REVIEW

**Comment by lirui-apache** [comment]
Added two more commits to fix `DataFormatConverters` and `ScalarFunctionCallGen`, as suggested by @JingsongLi

**Comment by JingsongLi** [comment]
LGTM

**Comment by lirui-apache** [comment]
cc @bowenli86 @zjuwangg @xuefuz

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarFunctionCallGen.scala [hunk 1]
```diff
@@ -79,7 +80,8 @@ class ScalarFunctionCallGen(scalarFunction: ScalarFunction) extends CallGenerato
         val boxedResultClass = ExtractionUtils.boxPrimitive(resultClass).asInstanceOf[Class[_]]
         val javaTypeTerm = boxedResultClass.getCanonicalName
         val resultExternalTypeWithResultClass =
-          if (resultExternalType.getLogicalType.supportsOutputConversion(boxedResultClass)) {
+          if (LogicalTypeDataTypeConverter.fromDataTypeToLogicalType(resultExternalType)
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/planner/codegen/calls/ScalarFunctionCallGen.scala]: Can you add a UDF case like:
```
class MyUdf extends ScalarFunction {
   public Object eval(...) {...}
   public TypeInformation getResultType() {
      return Types.BIG_DECIMAL;
   }
}
```
to repeat this bug?

File: flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java [hunk 2]
```diff
@@ -164,9 +164,10 @@ public static DataFormatConverter getConverterForDataType(DataType originDataTyp
 				Tuple2<Integer, Integer> ps = getPrecision(logicalType);
 				if (clazz == BigDecimal.class) {
 					return new BigDecimalConverter(ps.f0, ps.f1);
-				} else {
+				} else if (clazz == Decimal.class) {
 					return new DecimalConverter(ps.f0, ps.f1);
 				}
+				throw new RuntimeException("Not support class for DECIMAL: " + clazz);
```
Comment (wuchong) [inline flink-table/flink-table-runtime-blink/src/main/java/org/apache/flink/table/dataformat/DataFormatConverters.java]: Put it in the `else` branch? Could you help to update other types which have the same problem, e.g. `TIMESTAMP_WITHOUT_TIME_ZONE`, `TIMESTAMP_WITH_LOCAL_TIME_ZONE`, `TIMESTAMP_WITH_LOCAL_TIME_ZONE`...

**Comment by lirui-apache** [comment]
@JingsongLi @wuchong Thanks for your review. PR updated accordingly.

**Pull request merged**